File: 53. Maximum Subarray.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        if not nums:
            return
        if len(nums) == 1:
            return nums[0]
        curr_max, total_max = 0, -2**31
        for i in range(len(nums)):
            curr_max = max(curr_max+nums[i], nums[i])
            total_max = max(curr_max, total_max)
        return total_max

File: 53. Maximum Subarray.java
class Solution {
    /**
     * Greedy approach
     * Time O(n)
     * Space O(1)
     * @param nums
     * @return
     */
    public int maxSubArray(int[] nums) {
        int currSum=nums[0], totalSum=nums[0];
        for (int i=1;i<nums.length;i++) {
            currSum = Math.max(nums[i], currSum + nums[i]);
            totalSum = Math.max(totalSum, currSum);
        }
        return totalSum;
    }
}